src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,132 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elementData;
+    private int size;
+
+    public ArrayList() {
+        this.elementData = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        resizeWhenFull();
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        indexCheckForAdd(index);
+        resizeWhenFull();
+        System.arraycopy(elementData, index,
+                elementData, index + 1,
+                size - index);
+        elementData[index] = value;
+        size += 1;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (!indexCheckForExist(index, size)) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of range");
+        }
+        return (T)elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (!indexCheckForExist(index, size)) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of range");
+        }
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (!indexCheckForExist(index, size)) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of range");
+        }
+        final Object[] temp = elementData;
+        T oldValue = (T) elementData[index];
+        fastRemove(temp, index);
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        final int size = this.size;
+        int i;
+        boolean noElement = false;
+        if (element == null) {
+            for (i = 0; i < size; i++) {
+                if (elementData[i] == null) {
+                    break;
+                }
+            }
+        } else {
+            for (i = 0; i < size; i++) {
+                if (element.equals(elementData[i])) {
+                    break;
+                }
+                if (i == size - 1) {
+                    throw new NoSuchElementException("NO such element");
+                }
+            }
+        }
```

Comment:
simplify this code, you only iterate through array and find your T element. Please remember about KISS principe.